core/amber/src/main/scala/edu/uci/ics/texera/workflow/common/workflow/PhysicalPlan.scala

```diff
@@ -17,31 +17,38 @@ import scala.jdk.CollectionConverters.{IteratorHasAsScala, SetHasAsScala}
 
 object PhysicalPlan {
 
-  def apply(operatorList: Array[PhysicalOp], links: Array[PhysicalLink]): PhysicalPlan = {
-    new PhysicalPlan(operatorList.toSet, links.toSet)
-  }
-
   def apply(context: WorkflowContext, logicalPlan: LogicalPlan): PhysicalPlan = {
 
     var physicalPlan = PhysicalPlan(operators = Set.empty, links = Set.empty)
 
-    logicalPlan.operators.foreach(op => {
-      val subPlan =
-        op.getPhysicalPlan(
-          context.workflowId,
-          context.executionId
-        )
-      physicalPlan = physicalPlan.addSubPlan(subPlan)
-    })
-
-    // connect external links
-    logicalPlan.links.foreach(link => {
-      val fromOp = physicalPlan.getPhysicalOpForOutputPort(link.fromOpId, link.fromPortId)
-      val toOp = physicalPlan.getPhysicalOpForInputPort(link.toOpId, link.toPortId)
-      physicalPlan =
-        physicalPlan.addLink(PhysicalLink(fromOp.id, link.fromPortId, toOp.id, link.toPortId))
+    logicalPlan.getTopologicalOpIds.asScala.foreach(logicalOpId => {
+      val logicalOp = logicalPlan.getOperator(logicalOpId)
+      logicalOp.setContext(context)
+
+      val subPlan = logicalOp.getPhysicalPlan(context.workflowId, context.executionId)
+      subPlan
+        .topologicalIterator()
+        .map(subPlan.getOperator)
+        .foreach({ physicalOp =>
+          {
+            val externalLinks = logicalPlan
+              .getUpstreamLinks(logicalOp.operatorIdentifier)
+              .filter(link => physicalOp.inputPorts.contains(link.toPortId))
+              .map(link => {
+                val fromOp = physicalPlan.getPhysicalOpForOutputPort(link.fromOpId, link.fromPortId)
+                PhysicalLink(fromOp.id, link.fromPortId, physicalOp.id, link.toPortId)
+              })
+
+            val internalLinks = subPlan.getUpstreamPhysicalLinks(physicalOp.id)
+
+            // Add the operator and all links to the physical plan
+            physicalPlan = (externalLinks ++ internalLinks)
+              .foldLeft(physicalPlan.addOperator(physicalOp.propagateSchema())) { (plan, link) =>
```

Comment:
add more comments to this function. It's hard to understand.

Reply:
Or break it into sub-functions.

Reply:
updated